backend/de.metas.ui.web.base/src/main/java/de/metas/ui/web/window/model/Document.java

```diff
@@ -1355,23 +1356,32 @@ private void updateReadonlyAndPropagate(final ReasonSupplier reason)
 		getFields().forEach(documentField -> updateFieldReadOnlyAndCollect(documentField, reason));
 	}
 
+	@NonNull
 	private DocumentReadonly computeReadonly()
 	{
-		final boolean fieldsReadOnlyBasedOnDecorators = getEntityDescriptor().getDocumentDecorators() != null
-				&& getEntityDescriptor().getDocumentDecorators()
-				.stream()
-				.anyMatch(documentDecorator -> documentDecorator.isReadOnly(this));
-
-		final boolean fieldsReadOnly = fieldsReadOnlyBasedOnDecorators || computeDefaultFieldsReadOnly().booleanValue();
-
 		return DocumentReadonly.builder()
 				.parentActive(parentReadonly.isActive()).active(isActive())
 				.processed(parentReadonly.isProcessed() || isProcessed())
 				.processing(parentReadonly.isProcessing() || isProcessing())
-				.fieldsReadonly(fieldsReadOnly)
+				.fieldsReadonly(ExtendedMemorizingSupplier.of(this::computeFieldsReadOnly))
 				.build();
 	}
 
+	private boolean computeFieldsReadOnly()
+	{
+		final boolean isReadOnlyLogicTrue = computeDefaultFieldsReadOnly().booleanValue();
+
+		if (isReadOnlyLogicTrue)
+		{
+			return true;
+		}
+
+		return getEntityDescriptor()
+				.getDocumentDecorators()
+				.stream()
+				.anyMatch(documentDecorator -> documentDecorator.isReadOnly(this));
```

Comment:
calling 

> documentDecorator.isReadOnly(this)

in a cached method might not be the best approach.

Why not calling it `documentDecorator.isReadOnly(TableRecordReference...)` ?

Reply:
also, what shall happen in case the external reference become readonly after we compute this cached value?
do we have this case?

Reply:
well, the IDocumentDecorator is about a document + there might be other implementations where isReadOnly will be computed based on other Document properties...

Reply:
> well, the IDocumentDecorator is about a document + there might be other implementations where isReadOnly will be computed based on other Document properties...

OK, but if we want to consider that.... then we also need to consider invalidating the cache when those "other Document properties" were changed.... wdyt?

Reply:
hmmm.. I get what you are saying... but it's not really impacting at all the way it was initially set up, as :
  - "before" `de.metas.ui.web.window.model.DocumentReadonly#fieldsReadonly` was directly the computed value, so whenever a field's "readOnly" state was checked in `de.metas.ui.web.window.model.DocumentReadonly#computeFieldReadonly` the `de.metas.ui.web.window.model.DocumentReadonly#fieldsReadonly` was always the same, i.e. same as it was when the document was computed and stored in cache... 
  - "now" we are simply delaying the moment `de.metas.ui.web.window.model.DocumentReadonly#fieldsReadonly` gets computed

On another note, thinking about the scenario where the `externalReference.isReadOnlyInMetasfresh` flag changes from Y to N or N to Y:
 - I've looked a bit deeper, and I've seen that the DocumentReadonly property of a document can be updated (while the document is in cache) via `de.metas.ui.web.window.descriptor.DocumentEntityDescriptor#dependencies` && `de.metas.ui.web.window.model.Document#updateFieldsWhichDependsOn` but it doesn't really help us... as we depend on a whole different record so what we could do, is have a cache reset posted when the 'isReadOnly'  flag changes in external reference... ( the `CacheInvalidateRequest` could be posted only for the referenced record )